services/wfm/src/main/java/org/openkilda/wfm/topology/nbworker/fsm/FlowValidationFsm.java

```diff
@@ -0,0 +1,190 @@
+                .callMethod("receiveRules");
+        builder.internalTransition().within(RECEIVE_RULES).on(RULES_RECEIVED).callMethod("receivedRules");
+
+        builder.externalTransition().from(RECEIVE_RULES).to(FINISHED_WITH_ERROR).on(ERROR)
+                .callMethod(FINISHED_WITH_ERROR_METHOD_NAME);
+        builder.externalTransition().from(RECEIVE_RULES).to(VALIDATE_FLOW).on(NEXT)
+                .callMethod("validateFlow");
+
+        builder.externalTransition().from(VALIDATE_FLOW).to(FINISHED_WITH_ERROR).on(ERROR)
+                .callMethod(FINISHED_WITH_ERROR_METHOD_NAME);
+        builder.externalTransition().from(VALIDATE_FLOW).to(FINISHED).on(NEXT)
+                .callMethod(FINISHED_METHOD_NAME);
+
+        return builder;
+    }
+
+    public String getKey() {
+        return key;
+    }
+
+    protected void initialized(FlowValidationState from, FlowValidationState to,
+                               FlowValidationEvent event, Object context) {
+        log.info("Key: {}; FSM initialized", key);
+        flowId = request.getFlowId();
+        service = new FlowValidationService(persistenceManager);
+    }
+
+    protected void receiveRules(FlowValidationState from, FlowValidationState to,
+                                FlowValidationEvent event, Object context) {
+        log.info("Key: {}; Send commands to get rules on the switches", key);
+        try {
+            List<SwitchId> switchIds = service.getSwitchIdListByFlowId(flowId);
+            awaitingRequests = switchIds.size();
+            switchIds.forEach(switchId -> carrier.sendCommandToSpeakerWorker(key, switchId));
+
+        } catch (FlowNotFoundException e) {
+            log.error("Key: {}; Flow {} not found when sending commands to SpeakerWorkerBolt", key, flowId);
+            sendException(e.getMessage(), "Receiving rules operation in FlowValidationFsm", ErrorType.NOT_FOUND);
+        }
+    }
+
+    protected void receivedRules(FlowValidationState from, FlowValidationState to,
+                                 FlowValidationEvent event, Object context) {
+        SwitchFlowEntries switchFlowEntries = (SwitchFlowEntries) context;
+        log.info("Key: {}; Switch rules received for switch {}", key, switchFlowEntries.getSwitchId());
+        recievedRules.add(switchFlowEntries);
+        if (--awaitingRequests == 0) {
+            fire(NEXT);
+        }
+    }
+
+    protected void validateFlow(FlowValidationState from, FlowValidationState to,
+                                FlowValidationEvent event, Object context) {
+        try {
+            response = service.validateFlow(flowId, recievedRules);
+        } catch (FlowNotFoundException e) {
+            log.error("Key: {}; Flow {} not found during flow validation", key, flowId);
+            sendException(e.getMessage(), "Flow validation operation in FlowValidationFsm", ErrorType.NOT_FOUND);
+        } catch (Exception e) {
+            log.error("Key: {}; {}", key, e.getMessage());
```

Comment:
We lose here entire stacktrace. Let's pass an exception to logger with more specific message